### PR TITLE
Fix Regression: Renewal Info on My Plan page

### PR DESCRIPTION
**Pull request by kanishkdudeja**
**Note:** This is my first Calypso/React PR. Please excuse any silly or serious mistakes since I'm still learning.

This PR does some refactoring to fix an issue caused by: https://github.com/Automattic/wp-calypso/pull/38486/ (see: https://github.com/Automattic/wp-calypso/pull/38486/files#r370240351)

Basically, the new `ProductExpiration` React component assumes that renewals for subscriptions will always happen at the expiration date. However, we renew WordPress.com subscriptions ahead of their expiration. Due to this assumption, Calypso currently shows the wrong renewal date on the `My Plan` page. This PR aims to address this issue.

#### Changes proposed in this Pull Request

* Adds a new `renewDateMoment` property to the `ProductExpiration` component so that separate expiration and renewal dates can be passed to the component. The component's logic is also modified so as to support this change. Along with that, this PR also updates the documentation 

* The `getExpirationInfo()` function in the `PurchasesListing` component is now splitted into separate functions: `getExpirationInfoForPlan()` and `getExpirationInfoForPurchase` since the older function was called both for plan and purchase objects. Both objects have significant differences with the `renewDate/autoRenewDate` and the `isRefundable` properties, so it's better to have different functions operate on these different objects. Let me know if this approach doesn't seem right to you, I'm open to hearing suggestions and refactor it to make it better.

##### Initial Setup

* Check out this pull request on your local environment

##### Test Renewal Info on My Plan page for a WordPress.com subscription

* Create a new site on your WordPress.com account and buy a WordPress.com premium/personal plan subscription on that site.
 * Navigate to `http://calypso.localhost:3000/plans/my-plan/{Your site address}`. Calypso should show the `Renews on {DATE}` text on the `My Plan` block. This will typically be a month before the plan's expiration.
* Now, disable auto renew on the plan subscription from SA.
* Refresh the page and it should now show: `Expires on {DATE}` text. This should be the same date as the plan's expiration.

##### Test Renewal Info on My Plan page for Jetpack Backup

* Set up a fresh test Jetpack site on your WordPress.com account and buy Jetpack Backup for the site.
 * Navigate to `http://calypso.localhost:3000/plans/my-plan/{Your site address}`. Calypso should show the `Renews on {DATE}` text on the `My Plan` block. This will typically be the same date as the product's expiration.
* Now, disable auto renew on the plan subscription from SA.
* Refresh the page and it should now show: `Expires on {DATE}` text. This should be the same date as the plan's expiration.

##### Test Renewal Info on Plans page for Jetpack Backup

* Navigate to `http://calypso.localhost:3000/plans/{Your site address}`. Calypso should show the `Purchased on {DATE}` text for the purchase under the `Solutions` heading (beneath the `Plans` section).
* Edit the code in `client/components/product-expiration/index.tsx` to remove the `|| isRefundable` condition on line 38. This is needed so that we can test out the other conditions.
* Now, refresh the page. It should now show: `Renews on {DATE}`. This date should be equal to the product's expiration date.
* Now, disable auto renew on the Jetpack backup subscription from SA.
* Now, refresh the page. It should now show: `Expires on {DATE}`. This date should be equal to the product's expiration date.
* Now, revert the code change you made in file `client/components/product-expiration/index.tsx`.

### CODE REVIEW

**Review by chrismccluskey — APPROVED** [review]
Great looking PR, from what I can tell! I've added some comments but they are pretty minor. I've tested this and it works as you've described. I'm marking this as approved, but as this is my first Calypso review, you may want to wait until other reviewers can take a look before deploying, in case there's something I missed.

File: client/components/product-expiration/README.md [hunk 1]
```diff
@@ -13,6 +13,9 @@ This controls the formatting of the date format, which is passed to a localized
 #### `expiryDateMoment { moment }`
 This is the expiration date as a `moment` instance, and is required for display.
 
+#### `renewDateMoment { moment }`
+This is the date on which the product will be renewed, as a `moment` instance, used only when provided.
```
Comment (DavidRothstein) [inline client/components/product-expiration/README.md]: (minor) It might be more clear to say "will be auto-renewed" here.

File: client/components/product-expiration/README.md [hunk 2]
```diff
@@ -25,6 +28,7 @@ Optional purchase date, used only when `isRefundable` provided.
 const expiryDateMoment = moment( product.expiry );
 <ProductExpiration
 	expiryDateMoment={ expiryDateMoment }
+	renewDateMoment={ renewDateMoment }
```
Comment (DavidRothstein) [inline client/components/product-expiration/README.md]: (minor) You should probably add a line which sets `renewDateMoment` above (similar to the line that sets `expiryDateMoment`).

File: client/components/product-expiration/test/index.js [hunk 3]
```diff
@@ -36,7 +36,11 @@ describe( 'ProductExpiration', () => {
 	it( 'should return the renewal date in when the date is in the future', () => {
 		const date = moment( new Date( 2100, 10, 10 ) );
 		const wrapper = shallow(
-			<ProductExpiration expiryDateMoment={ date } translate={ translate } />
+			<ProductExpiration
+				expiryDateMoment={ date }
+				renewDateMoment={ date }
```
Comment (DavidRothstein) [inline client/components/product-expiration/test/index.js]: It could be useful to have this test a scenario where `expiryDateMoment` and `renewDateMoment` are different dates.

Separately, it might be nice to add one new test case to cover the new "Expires on..." scenario added in this PR.

File: client/components/product-expiration/index.tsx [hunk 4]
```diff
@@ -50,8 +52,12 @@ export class ProductExpiration extends React.PureComponent< Props > {
 			return translate( 'Expired on %s', { args: expiryDateMoment.format( dateFormat ) } );
 		}
 
+		if ( ! renewDateMoment ) {
```
[the remark this answers is not in the document]
Comment (DavidRothstein) [reply client/components/product-expiration/index.tsx]: I agree.

File: client/my-sites/plans/current-plan/purchases-listing.jsx [hunk 5]
```diff
@@ -124,7 +124,52 @@ class PurchasesListing extends Component {
...
+
+		const expiryMoment = plan.expiryDate ? this.props.moment( plan.expiryDate ) : null;
+
+		const renewMoment =
+			plan.autoRenew && plan.autoRenewDate ? this.props.moment( plan.autoRenewDate ) : null;
+
+		let isRefundable = false;
+
+		// we need to find the purchase linked to the current plan to find out if it's refundable or not
+		const purchaseForCurrentPlan = this.getPurchaseByCurrentPlan();
```
Comment (DavidRothstein) [inline client/my-sites/plans/current-plan/purchases-listing.jsx]: Technically you're fixing a bug here, but in practice I think it makes the UI worse.

If you have either a WordPress.com plan or a Jetpack plan, then before this diff you'll see "Renews on [Date]" when you go to `plans/my-plan/[site]`.  After this diff, you'll see "Purchased on [Date]" there.  I don't think that's a good change.  The user is already on a page that is designed to show their current purchases, so they don't need extra confirmation specifically telling them that they purchased it.

I think we should actually go the other direction and stop showing "Purchased on [Date]" for Jetpack Backup on this page too, to make it consistent with the others.

I tried to figure out how this change got made in the first place and I think I see what happened:
1. In the threaded discussion underneath p8oabR-qW-p2#comment-3622 there was a discussion of conditionally showing "Purchased on [Date]" for Jetpack Backup on the *actual* plans page (`plans/[site]`), and it probably does make some sense there.  (That is the page you go to in order to buy new things, so if you just bought Jetpack Backup recently it is helpful to see that specifically.)
2. Somehow when it was implemented, though, it got added to both `plans/[site]` and `plans/my-plan/[site]`.  I think that was a mistake.

Bottom line: My suggestion is that for the two usages of the `<ProductExpiration>` component in this file, we shouldn't pass in either `purchaseDateMoment` or `isRefundable` at all.  That would have the desired effect of showing only the "Renews"/"Expires" text at  `plans/my-plan/[site]`, while still conditionally showing the "Purchased" text at `plans/[site]`.

If that makes sense, I'd be happy to leave a comment on the above P2 thread explaining my reasoning and with a link back here.

File: client/my-sites/plans/current-plan/purchases-listing.jsx [hunk 6]
```diff
@@ -124,7 +124,52 @@ class PurchasesListing extends Component {
...
+
+		return find(
+			purchases,
+			purchase => purchase.active && purchase.productSlug === currentPlanSlug
+		);
+	}
+
+	getExpirationInfoForPlan( plan ) {
+		// No expiration date for free plan or partner site.
+		if ( this.isFreePlan( plan ) || isPartnerPurchase( plan ) ) {
```
Comment (DavidRothstein) [inline client/my-sites/plans/current-plan/purchases-listing.jsx]: It seems wrong to run a check called on `isPartnerPurchase()` on something that isn't a purchase.. I am not sure, but I suspect this code never worked before.

Similarly, `this.isFreePlan()` is called here on a plan, but further down it's called on a purchase... looking at the code, that might actually work in practice though.

Bottom line: I think there are existing bugs here (that you didn't cause) and I'm not sure the best way to handle them; they just come into glaring relief when you split this up into separate functions for the plan and purchase expiration :)

**Review by DavidRothstein — CHANGES_REQUESTED** [review]
Overall looks good to me also.  But I found one major behavior change that this leads to, which I don't think is intentional (more details in an inline comment) and I left a few other inline comments also.

>The getExpirationInfo() function in the PurchasesListing component is now splitted into separate functions... Let me know if this approach doesn't seem right to you, I'm open to hearing suggestions and refactor it to make it better.

If you really wanted to get into refactoring it, one option would be to use the `getPurchaseByCurrentPlan()` method that you added here to convert the plan to a purchase before passing it in -- in that case, you would only need one function for getting the expiration info (and it would only have to work with a purchase object).

But I'm not sure that's worth the effort, especially given that it could break something else... So I'm also happy with the way you did it, since it seems to get the job done practically.

File: client/my-sites/plans/current-plan/purchases-listing.jsx [hunk 7]
```diff
@@ -124,7 +124,52 @@ class PurchasesListing extends Component {
...
+	getPurchaseByCurrentPlan() {
+		const { currentPlanSlug, purchases } = this.props;
+
+		if ( ! currentPlanSlug ) {
+			return null;
+		}
+
+		return find(
+			purchases,
+			purchase => purchase.active && purchase.productSlug === currentPlanSlug
```
[the remark this answers is not in the document]
Comment (kanishkdudeja) [reply client/my-sites/plans/current-plan/purchases-listing.jsx]: Good suggestion @chrismccluskey. However, after feedback from Chris, it looks like we don't need this function anyway. So, I've removed this function and it's usage.

File: client/my-sites/plans/current-plan/purchases-listing.jsx [hunk 8]
```diff
@@ -124,7 +124,52 @@ class PurchasesListing extends Component {
...
+		const renewMoment =
+			plan.autoRenew && plan.autoRenewDate ? this.props.moment( plan.autoRenewDate ) : null;
+
+		let isRefundable = false;
+
+		// we need to find the purchase linked to the current plan to find out if it's refundable or not
+		const purchaseForCurrentPlan = this.getPurchaseByCurrentPlan();
+
+		if ( purchaseForCurrentPlan ) {
+			isRefundable = purchaseForCurrentPlan.isRefundable;
```
[the remark this answers is not in the document]
Comment (kanishkdudeja) [reply client/my-sites/plans/current-plan/purchases-listing.jsx]: Thanks @chrismccluskey. It looks like we don't need this code block now (after taking into account David's feedback), so this code has now been removed.

File: client/components/product-expiration/index.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (kanishkdudeja) [reply client/components/product-expiration/index.tsx]: Great suggestion. This is implemented in the latest commit.

File: client/components/product-expiration/README.md (hunk 1 above)
Comment (kanishkdudeja) [reply client/components/product-expiration/README.md]: Thanks. Implemented.

File: client/components/product-expiration/README.md (hunk 2 above)
Comment (kanishkdudeja) [reply client/components/product-expiration/README.md]: Thanks. Implemented.

File: client/components/product-expiration/test/index.js (hunk 3 above)
Comment (kanishkdudeja) [reply client/components/product-expiration/test/index.js]: Thanks. Implemented.

File: client/my-sites/plans/current-plan/purchases-listing.jsx (hunk 5 above)
Comment (kanishkdudeja) [reply client/my-sites/plans/current-plan/purchases-listing.jsx]: Thanks. Implemented.

File: client/my-sites/plans/current-plan/purchases-listing.jsx (hunk 6 above)
Comment (kanishkdudeja) [reply client/my-sites/plans/current-plan/purchases-listing.jsx]: I was a little worried about this too. I've removed the `isPartnerPurchase()` check for plans but have still kept the the `isFreePlan()` check for purchases since as you said, that might still work in practice.

**Comment by kanishkdudeja** [comment]
Thanks for your feedback @chrismccluskey and @DavidRothstein. I've made changes as per your feedback.

File: client/blocks/product-selector/index.jsx [hunk 9]
```diff
@@ -192,9 +192,12 @@ export class ProductSelector extends Component {
 
 		const expiryMoment = purchase.expiryDate ? moment( purchase.expiryDate ) : null;
 
+		const renewDateMoment = purchase.renewDate ? moment( purchase.renewDate ) : null;
```
Comment (DavidRothstein) [inline client/blocks/product-selector/index.jsx]: Should this actually be something more like:
```
const renewDateMoment = ! isExpiring( purchase ) && purchase.renewDate ? moment( purchase.renewDate ) : null;
```

So that it actually checks the auto-renew status?

The `isExpiring()` check is what is used elsewhere (e.g. on the `/me/purchases` page) so this would ensure consistency with that.

While testing, I actually did manage to stumble upon a scenario where the `/me/purchases` page said auto-renew was off but the code here (incorrectly) acted like it was on.  I could not figure out how to reproduce it again (nor was it obvious to me how it happened, after looking at the server-side code), but somehow it did, so I guess it's worth protecting against.

File: client/my-sites/plans/current-plan/purchases-listing.jsx [hunk 10]
```diff
@@ -124,25 +124,31 @@ class PurchasesListing extends Component {
...
 		const expiryMoment = purchase.expiryDate ? this.props.moment( purchase.expiryDate ) : null;
 
-		return (
-			<ProductExpiration
-				expiryDateMoment={ expiryMoment }
-				purchaseDateMoment={ subscribedMoment }
-				isRefundable={ purchase.isRefundable }
-			/>
-		);
+		const renewMoment = purchase.renewDate ? this.props.moment( purchase.renewDate ) : null;
```
Comment (DavidRothstein) [inline client/my-sites/plans/current-plan/purchases-listing.jsx]: Similar to my above comment, we may want to add an `isExpiring()` check here.

**Review by DavidRothstein — COMMENTED** [review]
Seems to mostly work and it all looks good.  I found one possible issue though. After that, it's probably ready.

File: client/my-sites/plans/current-plan/purchases-listing.jsx (hunk 10 above)
Comment (kanishkdudeja) [reply client/my-sites/plans/current-plan/purchases-listing.jsx]: Thanks. Added it here as well.

File: client/blocks/product-selector/index.jsx (hunk 9 above)
Comment (kanishkdudeja) [reply client/blocks/product-selector/index.jsx]: Thanks @DavidRothstein. I didn't encounter this edge case while testing and assumed that `renewDate` sent from the backend will be empty if the auto-renew is off.

But it's better to be safe, so I've added this check.

**Comment by kanishkdudeja** [comment]
Thanks for your feedback @DavidRothstein. I've made changes as per your feedback and this looks good to merge now.

**Comment by gwwar** [comment]
@kanishkdudeja let us know if you need help with rebasing with master

**Review by gwwar — DISMISSED** [review]
Feel free to dismiss this review after stray commits are no longer in this branch 👍
